### PR TITLE
ByteString introduced as AsciiString super class

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The usage and code within AsciiString has exceeded the original design scope for this class. Its usage as a binary string is confusing and on the verge of violating interface assumptions in some spots.

Modifications:
- ByteString will be created as a base class to AsciiString. All of the generic byte handling processing will live in ByteString and all the special character encoding will live in AsciiString.

Results:
The AsciiString interface will be clarified. Users of AsciiString can now be clear of the limitations the class imposes while users of the ByteString class don't have to live with those limitations.

### CODE REVIEW

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestEncoder.java [hunk 1]
```diff
@@ -38,7 +39,7 @@ public boolean acceptOutboundMessage(Object msg) throws Exception {
     @Override
     protected void encodeInitialLine(ByteBuf buf, HttpRequest request) throws Exception {
         AsciiString method = request.method().name();
-        buf.writeBytes(method.array(), method.arrayOffset(), method.length());
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestEncoder.java]: @trustin @normanmaurer - Note that I got rid of the `arrayOffset()` method to simplify the interface a bit. I don't think this is a problem if when creating a `ByteString` a copy operation is done....however if the intention is to not copy (share the underlying array) we won't be able to do that effectively without `arrayOffset()`.

Should I restore `arrayOffset()` and allow "no copy of sub array" operations to take place?  I can submit a followup PR perhaps...I think a few areas will need investigation if we support `arrayOffset()`.

**Comment by Scottmitch** [comment]
@trustin @normanmaurer - This is for forward porting https://github.com/netty/netty/pull/3579 and I created a PR because I have a question.  Also it is a relatively big merge so feel free to review in general...

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestEncoder.java (hunk 1 above)
Comment (trustin) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestEncoder.java]: I left `arrayOffset()` just in case `AsciiString`/`ByteString` wraps a subregion of an existing byte array, which is not the case currently.  Non-zero `arrayOffset()` introduces more complexity in indexing and we are unsure about its performance impact.

Maybe we could remove `arrayOffset()` and let a user assume the array returned by `array()` will always have its first character at 0th index and have the exact same length with the `ByteString`, but I felt it might limit our chances to explore the potential room for improvements in the future.

File: common/src/main/java/io/netty/util/AsciiString.java [hunk 2]
```diff
@@ -1331,155 +1016,70 @@ public boolean contains(CharSequence cs) {
...
-        return Float.parseFloat(toString(start, end));
+        return parseAsciiFloat(start, end);
     }
 
     public double parseDouble() {
-        return parseDouble(0, length());
+        return parseAsciiDouble();
     }
 
     public double parseDouble(int start, int end) {
```
Comment (trustin) [inline common/src/main/java/io/netty/util/AsciiString.java]: This will make IDE show `parseDouble()` and `parseAsciiDouble()` which do the same job. Could we just remove the ones at `AsciiString`? Please also port this change to 4.1 if we have the same issue there.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestEncoder.java (hunk 1 above)
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestEncoder.java]: I agree with @trustin here. Better safe than sorry ;)

Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestEncoder.java]: @trustin Agreed it introduced more complexity...but allows for us to wrap subregions as I discussed.  I think we are in agreement that I should submit a follow-up PR to put `arrayOffset()` back in?

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 2 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: Yip good catch.  Forgot to clean these up.

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: I will pull this in and then submit a followup PR for this and `arrayOffset()`.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/AsciiString.java]: @Scottmitch +1 pull in and do another PR. Great work as usual !

**Comment by Scottmitch** [comment]
Cherry-picked in to master (b823bfa)

Will submit a followup PR for 4.1 and master to add `arrayOffset()` and cleanup the `parse*` routines.
